Yoast/wordpress-seo
# Bf plugin ad 4 sidebar metabox

## Context


* We want to show notifications related to the Black Friday sales during the Black Friday week, both in sidebar and metabox.
 
## Summary


This PR can be summarized in the following changelog entry:

*  Adds a Black Friday notification in editors sidebar and metabox.

## Relevant technical choices:

* The components and the logic introduced in https://github.com/Yoast/wordpress-seo/pull/20615 has been refactored to use the same logic everywhere.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Preliminary steps
* Test this PR on a website which has an unauthorizer Yoast SEO Premium
* Deactivate Yoast SEO Premium
* Make sure you have no notifications alredy dismissed
  * With a MySQL client edit the `wp_usermeta` table
  * search  for the tuple with `user_id` = your user id and `meta_key` = `_yoast_alerts_dismissed`
    * set `meta_value` to `a:0:{}`  
* Edit `src/promotions/domain/black-friday-promotion.php`
  * change `\gmmktime( 11, 00, 00, 11, 23, 2023 )` to `\gmmktime( 11, 00, 00, 11, 23, 2022 )`


#### Test the sidebar notice
* Edit a post
* In the Yoast sidebar, check that you see the following notification:
  <img width="303" alt="Screenshot 2023-09-12 at 12 46 51" src="https://github.com/Yoast/wordpress-seo/assets/68744851/bab35046-57e9-4fb7-9b45-7ce90669e9d2">
* Dismiss the notification
* Refresh the page and verify you now get the following notification:
  <img width="300" alt="Screenshot 2023-09-12 at 12 48 15" src="https://github.com/Yoast/wordpress-seo/assets/68744851/5a188003-1219-4d22-b4f3-914a9eca1392">
* Dismissis this notification, too
* Refresh the page and verify no notifications are shown


#### Test the metabox notice
* Follow the instructions in the *Preliminary steps* section to reset your notification status
* Edit a post
* In the metabox, verify you see the following notification:
  <img width="632" alt="Screenshot 2023-09-12 at 12 52 17" src="https://github.com/Yoast/wordpress-seo/assets/68744851/236f877d-152d-4f67-8abb-3329f7a7af2a">
* Dismiss the notification and refresh the page
* Verify the notification is gone both from the metabox and the sidebar (in the sidebar you should see the free webinar notification instead)


#### Test the notice is not shown when a warning is present in the metabox
* Follow the instructions in the *Preliminary steps* section to reset your notification status
* Activate  Yoast SEO Premium
* Edit a post
* Verify that both in the metabox and the sidebar you get a warning instead of the Black Friday notification
  * metabox warning: 
    <img width="628" alt="Screenshot 2023-09-05 at 00 33 29" src="https://github.com/Yoast/wordpress-seo/assets/68744851/5f6146cd-5414-40e7-b305-d94dfa303d9f">
  * sidebar warning:
    <img width="298" alt="Screenshot 2023-09-12 at 12 59 11" src="https://github.com/Yoast/wordpress-seo/assets/68744851/7f3e0d20-e126-464f-9e9c-62e0c4c965b9">


#### Test the notice is not shown outside its time frame
* Follow the instructions in the *Preliminary steps* section to reset your notification status
* De-activate Premium
* Revert the changes you did to `src/promotions/domain/black-friday-promotion.php`
* Verify that no Black Friday notification is show neither in the metabox or sidebar (in the sidebar you should see the free webinar notification instead)

#### Relevant test scenarios


### Test instructions for QA when the code is in the RC



## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

* Please also perform a regression test on https://github.com/Yoast/wordpress-seo/pull/20615, as the logic has been rewritten to be the same everywhere in the code. 

## UI changes


## Other environments


## Documentation


## Quality assurance


## Innovation


Fixes https://github.com/Yoast/ux/issues/37
